### PR TITLE
Improve test coverage and docstrings for `generators/classic`

### PR DESCRIPTION
**Pull request by Peiffap**
Adds various additional tests (and improve existing ones) to take test coverage to 100 % for `generators/classic`.

Clean up the docstrings of the affected functions along the way.

### CODE REVIEW

**Review by rossbar — COMMENTED** [review]
Thanks for going through this @Peiffap !

One takeaway I have from looking through these changes is that I think it's important to be cognizant of over-parametrization. While it is indeed important to thoroughly test, it's also important to be aware of the implications to the test suite runtime! Test parametrization is a very powerful tool (you'll find quite a few `pytest.mark.parametrize` lines in my git blame), but it also makes it quite easy to over-specify tests. A simple statement like:

```python
@pytest.mark.parametrize("x", range(10))
@pytest.mark.parametrize("y", range(10))
@pytest.mark.parametrize("z", range(10))
```
looks harmless enough... but results in 1000 new test cases! Now, if each one of those cases represents a unique code path or some form of semantic corner case, then IMO it is indeed absolutely worthwhile to test each one individually. However more often this is not the case.

There is of course, no "right answer" to how exhaustive tests should be, but I wanted to share my impression from going through these! I've left a couple inline comments/suggestions with further details in specific instances.

File: networkx/generators/classic.py [hunk 1]
```diff
@@ -212,39 +222,49 @@ def barbell_graph(m1, m2, create_using=None):
...
-        Size of the left and right barbells, must be greater than 2.
+        Size of the left and right barbells.
 
     m2 : int
         Length of the path connecting the barbells.
 
     create_using : NetworkX graph constructor, optional (default=nx.Graph)
-       Graph type to create. If graph instance, then cleared before populated.
-       Only undirected Graphs are supported.
+        Graph type to create. If graph instance, then cleared before populated.
```
[the remark this answers is not in the document]
Comment (Peiffap) [reply networkx/generators/classic.py:231]: I think there is value in keeping it, but I'm worried about having two sources of truth. I'm not opposed to keeping it; I just thought I'd give my reasons for removing.

File: networkx/generators/classic.py [hunk 2]
```diff
@@ -891,15 +930,15 @@ def trivial_graph(create_using=None):
...
-    The Turan Graph is a complete multipartite graph on $n$ nodes
+    The Turán graph is a complete multipartite graph on $n$ nodes
     with $r$ disjoint subsets. That is, edges connect each node to
     every node not in its subset.
 
     Given $n$ and $r$, we create a complete multipartite graph with
-    $r-(n \mod r)$ partitions of size $n/r$, rounded down, and
-    $n \mod r$ partitions of size $n/r+1$, rounded down.
+    $r-(n \mod r)$ partitions of size $\lfloor n/r \rfloor$, and
+    $n \mod r$ partitions of size $\lfloor n/r \rfloor + 1$.
```
[the remark this answers is not in the document]
Comment (Peiffap) [reply networkx/generators/classic.py:941]: I've probably read more raw TeX math than the vast majority of people, so this is a definite blind spot for me! I try to be careful about introducing TeX, especially in docstrings that don't already contain math-mode expressions. I thought it made sense here to add it because 1. the previous expression was technically wrong and 2. the floor commands are fairly readable in this particular situation, in my opinion.

File: networkx/generators/tests/test_classic.py [hunk 3]
```diff
@@ -355,30 +358,85 @@ def test_lollipop_graph_size_node_sequence(self, m, n):
...
-    def test_lollipop_graph_exceptions(self):
-        # Raise NetworkXError if m<2
-        pytest.raises(nx.NetworkXError, nx.lollipop_graph, -1, 2)
-        pytest.raises(nx.NetworkXError, nx.lollipop_graph, 1, 20)
-        pytest.raises(nx.NetworkXError, nx.lollipop_graph, "", 20)
-        pytest.raises(nx.NetworkXError, nx.lollipop_graph, "a", 20)
-
-        # Raise NetworkXError if n<0
-        pytest.raises(nx.NetworkXError, nx.lollipop_graph, 5, -2)
+    @pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (Peiffap) [reply networkx/generators/tests/test_classic.py:361]: That does sound really interesting to me! Thanks for suggesting it. It's not entirely clear to me what you mean with it though (e.g. do you imagine this being a custom NetworkX solution, or a more general, existing framework applied to our codebase?), and it's definitely beyond the scope of this particular PR. Perhaps something for a separate issue? Do you know of any other places this has been discussed on NetworkX that I can pore over when I have some time? A quick search suggests the `hypothesis` library is a popular tool for this type of testing, and I'd be prepared to give that a go if given the green light.

I'm leaving these tests as they are for now, but if you want me to revert or reduce them in this PR, I can do that too.

File: networkx/generators/tests/test_classic.py [hunk 4]
```diff
@@ -355,30 +358,85 @@ def test_lollipop_graph_size_node_sequence(self, m, n):
...
+        ("create_using", "expectation"),
+        [
+            (nx.Graph, does_not_raise()),  # ok
+            (nx.Graph([(0, 1)]), does_not_raise()),  # ok
+            (nx.DiGraph, pytest.raises(nx.NetworkXError)),  # directed
+            (nx.DiGraph([(0, 1)]), pytest.raises(nx.NetworkXError)),  # directed
+            (nx.MultiGraph, does_not_raise()),  # ok
+            (nx.MultiGraph([(0, 1)]), does_not_raise()),  # ok
+            (nx.MultiDiGraph, pytest.raises(nx.NetworkXError)),  # directed
+            (nx.MultiDiGraph([(0, 1)]), pytest.raises(nx.NetworkXError)),  # directed
```
[the remark this answers is not in the document]
Comment (Peiffap) [reply networkx/generators/tests/test_classic.py:421]: I see. In my personal opinion, it doesn't make sense to allow something but discourage it (implictly? I didn't see it mentioned anywhere, but I might be missing something). I know this would _majorly_ suck to deprecate everywhere `create_using` is used, but I can't come up with another "good" reason to handle things this way.

**Comment by Peiffap** [comment]
Hey @rossbar, thanks for looking at this PR and for your insightful comments. I've committed some of the suggested changes, and responded directly to some others.

When it comes your overarching comment on overparametrization, I definitely see what you mean... I might've gone a little overboard (it is a _really_ nice feature...). Like you mention, it makes sense to cover all potential code paths, or semantically distinct cases. I was aware of this when I made the PR originally, but I had several reasons for keeping the overly exhaustive parametrized tests:
 * The overparametrization is actually what allowed me to find the error in the Turán graph docstring - some of the tests raised errors when they shouldn't have.
 * I've seen other PRs being held back because changes to tests made it hard to see whether the original tests were still passing - I wanted to get ahead of that and make sure I only _added_ test cases, without removing existing ones.
 * I was actually mindful of test suite runtime, and while your general comment holds (there are definitely cases where this is a concern), I think even at its worst (when I accidentally tested thousands of possible permutations for one of the complete multipartite tests), it was only adding a few seconds to runtime, mainly during test collection.

I want to emphasize this comment is not meant to disparage your comments! I appreciate you sharing your thoughts and agree with your assessment; I just wanted to provide context for some of the choices. Feel free to let me know what further changes you deem necessary or opportune!
